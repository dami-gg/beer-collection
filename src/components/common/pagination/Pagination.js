// @flow
import React, { PureComponent } from "react";

import PageButton from "./PageButton";

import { getButtonLabels } from "./pagination.helpers";
import { PLACEHOLDER_PAGE_LABEL } from "./pagination.constants";

import "./pagination.scss";

type Props = {
  className: string,
  numItems: number,
  currentPage: number,
  resultsPerPage: number,
  totalPageButtons: number,
  onNavigation: Function
};

class Pagination extends PureComponent<Props> {
  totalPages: number;
  navigateToPage: Function;
  navigateToPreviousPage: Function;
  navigateToNextPage: Function;

  constructor(props: Props) {
    super(props);

    this.navigateToPage = this.navigateToPage.bind(this);
    this.navigateToPreviousPage = this.navigateToPreviousPage.bind(this);
    this.navigateToNextPage = this.navigateToNextPage.bind(this);

    this.totalPages = this.getTotalPages();
  }

  componentWillUpdate(nextProps: Object) {
    this.totalPages = this.getTotalPages(nextProps.numItems);
  }

  getTotalPages(numItems: number = this.props.numItems) {
    return Math.ceil(numItems / this.props.resultsPerPage);
  }

  navigateToPage(page: number) {
    if (page !== PLACEHOLDER_PAGE_LABEL) {
      this.props.onNavigation(page);
    }
  }

  navigateToPreviousPage() {
    if (this.props.currentPage !== 1) {
      this.navigateToPage(this.props.currentPage - 1);
    }
  }

  navigateToNextPage() {
    if (this.props.currentPage !== this.totalPages) {
      this.navigateToPage(this.props.currentPage + 1);
    }
  }

  generatePageButtons() {
    return getButtonLabels(
      this.totalPages,
      this.props.totalPageButtons,
      this.props.currentPage
    ).map((item, i) => (
      <PageButton
        key={i}
        label={item}
        className={
          Math.abs(i - (this.props.currentPage - 1)) > 1
            ? "small-screens--hidden"
            : ""
        }
        clickHandler={this.navigateToPage}
        active={this.props.currentPage === item}
        disabled={item === PLACEHOLDER_PAGE_LABEL}
      />
    ));
  }

  generateOutmostButton(position: string = "left") {
    if (this.props.numItems && this.props.numItems > 0) {
      const isLeftPositioned = position === "left";
      const isDisabled = isLeftPositioned
        ? this.props.currentPage === 1
        : this.props.currentPage === this.totalPages;

      return (
        <PageButton
          label={
            isLeftPositioned ? (
              <i className={`fa fa-arrow-left`} aria-hidden="true" />
            ) : (
              <i className={`fa fa-arrow-right`} aria-hidden="true" />
            )
          }
          disabled={isDisabled}
          clickHandler={
            isLeftPositioned
              ? this.navigateToPreviousPage
              : this.navigateToNextPage
          }
        />
      );
    }
  }

  render() {
    return (
      <div
        className={`pagination ${this.totalPages === 1
          ? "pagination--hidden"
          : ""} ${this.props.className}`}>
        {this.generateOutmostButton("left")}

        {this.generatePageButtons()}

        {this.generateOutmostButton("right")}
      </div>
    );
  }
}

export default Pagination;
