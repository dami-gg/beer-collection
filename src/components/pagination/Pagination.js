// @flow
import type {Beer } from '../../types';

import React, { Component } from 'react';
import { connect } from 'react-redux';

import { RESULTS_PER_PAGE } from '../../constants';

class Pagination extends Component {
    props: {
        collection: Array<Beer>,
        currentPage: number
    };

    getTotalPages() {
        return this.props.collection ? (this.props.collection.length / RESULTS_PER_PAGE) + 1 : 0;
    }

    getPagination(currentPage: number) {
        let totalPages = 3; // TODO this.getTotalPages();

        return [...Array(totalPages)].map((item, i) =>
            `<li className="${currentPage === i ? 'active' : ''}" key="${i}">${i}</li>`
        );
    }

    render() {
        return (
            <ul className="pagination">
                {
                    this.getTotalPages() > 1 &&
                    this.getPagination(this.props.currentPage)
                }
            </ul>
        );
    }
}

const mapStateToProps = (state: Object) => ({
    collection: state.collection,
    currentPage: state.currentPage || 1 // TODO
});

export default connect(mapStateToProps)(Pagination);