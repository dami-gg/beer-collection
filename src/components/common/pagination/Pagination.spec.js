import React from "react";
import Pagination from "./Pagination";

import { shallow } from "enzyme";
import renderer from "react-test-renderer";

describe("Pagination", () => {
  let component, element, mock;

  beforeEach(mocks);

  it("should render without crashing", () => {
    shallow(<Pagination />);
  });

  it("should match the snapshot", () => {
    component = renderer.create(<Pagination />);
    let json = component.toJSON();

    expect(json).toMatchSnapshot();
  });

  describe("render", () => {
    it("should be hidden when there is only one page", () => {
      mock.props.numItems = 5;
      mock.props.resultsPerPage = 10;

      render();

      element = component.find(".pagination--hidden");

      expect(element.length).toEqual(1);
    });
  });

  function render() {
    component = shallow(
      <Pagination
        className={mock.props.className}
        numItems={mock.props.numItems}
        currentPage={mock.props.currentPage}
        resultsPerPage={mock.props.resultsPerPage}
        totalPageButtons={mock.props.totalPageButtons}
        onNavigation={mock.props.onNavigation}
      />
    );
  }

  function mocks() {
    mock = {
      context: {},
      props: {
        className: "",
        numItems: 5,
        currentPage: 1,
        resultsPerPage: 10,
        totalPageButtons: 1,
        onNavigation: jest.fn()
      }
    };
  }
});
