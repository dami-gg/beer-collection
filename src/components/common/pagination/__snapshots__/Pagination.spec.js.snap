// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should match the snapshot 1`] = `
<div
  className="pagination  undefined"
>
  <a
    className="pagination__button
                     
                     "
    onClick={[Function]}
  >
    <span>
      «
    </span>
  </a>
  <a
    className="pagination__button
                     
                     "
    onClick={[Function]}
  >
    <span>
      1
    </span>
  </a>
  <a
    className="pagination__button
                     
                     pagination__button--disabled"
    onClick={[Function]}
  >
    <span>
      ...
    </span>
  </a>
  <a
    className="pagination__button
                     
                     "
    onClick={[Function]}
  >
    <span>
      NaN
    </span>
  </a>
  <a
    className="pagination__button
                     pagination__button--active
                     "
    onClick={[Function]}
  >
    <span />
  </a>
  <a
    className="pagination__button
                     
                     "
    onClick={[Function]}
  >
    <span>
      NaN
    </span>
  </a>
  <a
    className="pagination__button
                     
                     pagination__button--disabled"
    onClick={[Function]}
  >
    <span>
      ...
    </span>
  </a>
  <a
    className="pagination__button
                     
                     "
    onClick={[Function]}
  >
    <span>
      NaN
    </span>
  </a>
  <a
    className="pagination__button
                     
                     "
    onClick={[Function]}
  >
    <span>
      »
    </span>
  </a>
</div>
`;
